`DataFrame` indexed with a generator expression (rather than list comprehension) is inferred as `Series`

**To Reproduce**
1. Provide a minimal runnable `pandas` example that is not properly checked by the stubs.
```python
from typing import TYPE_CHECKING

import pandas as pd

df = pd.DataFrame()
df["Col0"] = [0, 0, 1]
df["Col1"] = [0, 1, 1]
df["Col2"] = [1, 1, 1]

# bug
sub_df_gen = df[("Col" + str(i) for i in range(2))]
print(type(sub_df_gen))  # <class 'pandas.core.frame.DataFrame'>
if TYPE_CHECKING:
    reveal_type(sub_df_gen)  # Revealed type is "pandas.core.series.Series[Any]
print(sub_df_gen)
print(sub_df_gen.any(axis=1))
print(sub_df_gen.all(axis=1))

# workaround
sub_df_list = df[["Col" + str(i) for i in range(2)]]
print(type(sub_df_list))  # <class 'pandas.core.frame.DataFrame'>
if TYPE_CHECKING:
    reveal_type(sub_df_list)  # Revealed type is "pandas.core.frame.DataFrame"
print(sub_df_list)
print(sub_df_list.any(axis=1))
print(sub_df_list.all(axis=1))

```

2. Indicate which type checker you are using (`mypy` or  `pyright`).
`mypy`

3. Show the error message received from that type checker while checking your example.
```
bug.py:14: note: Revealed type is "pandas.core.series.Series[Any]"
bug.py:16: error: Argument "axis" to "any" of "Series" has incompatible type "Literal[1]"; expected "Literal['index', 0]"  [arg-type]
bug.py:17: error: Argument "axis" to "all" of "Series" has incompatible type "Literal[1]"; expected "Literal['index', 0]"  [arg-type]
bug.py:23: note: Revealed type is "pandas.core.frame.DataFrame"
Found 2 errors in 1 file (checked 1 source file)
```

**Please complete the following information:**
 - OS:Windows 11
 - OS Version: 22H2
 - python version: 3.11.10
 - version of type checker: 1.2.0
 - version of installed `pandas-stubs`: 2.0.1.230501

--- a/pandas-stubs/core/frame.pyi
+++ b/pandas-stubs/core/frame.pyi
@@ -1,5 +1,6 @@
 from collections.abc import (
     Callable,
+    Generator,
     Hashable,
     Iterable,
     Iterator,
@@ -550,14 +551,15 @@ class DataFrame(NDFrame, OpsMixin):
     def T(self) -> DataFrame: ...
     def __getattr__(self, name: str) -> Series: ...
     @overload
-    def __getitem__(
+    def __getitem__(  # type: ignore[misc]
         self,
         key: Series[_bool]
         | DataFrame
         | Index
         | np_ndarray_str
         | np_ndarray_bool
-        | list[_ScalarOrTupleT],
+        | list[_ScalarOrTupleT]
+        | Generator[_ScalarOrTupleT, None, None],
     ) -> DataFrame: ...
     @overload
     def __getitem__(self, key: slice) -> DataFrame: ...

--- a/tests/test_frame.py
+++ b/tests/test_frame.py
@@ -2715,3 +2715,8 @@ def test_groupby_fillna_inplace() -> None:
     check(assert_type(groupby.fillna(0, inplace=False), pd.DataFrame), pd.DataFrame)
     if TYPE_CHECKING_INVALID_USAGE:
         groupby.fillna(0, inplace=True)  # type: ignore[arg-type] # pyright: ignore[reportGeneralTypeIssues]
+
+
+def test_getitem_generator() -> None:
+    # GH 685
+    check(assert_type(DF[(f"col{i+1}" for i in range(2))], pd.DataFrame), pd.DataFrame)

